Implementation for `@unknown default`

jrose-apple opened the pull request:
Part of the [revised SE-0192](https://github.com/apple/swift-evolution/pull/777), building on top of #11961. The first PR is big enough that I'd like to land this part separately afterwards.

~Still to do: write some SIL tests to make sure `unknown case` is properly getting treated like `default`.~ (done!)

jrose-apple committed:
Diagnose when switching on a non-frozen enum without a catch-all case

The first half of Sema support for '@unknown'. The other part is
handling when the user /does/ write '@unknown', which results in
/other/ things being downgraded to warnings.

The diagnostics here are still pretty minimal; they should explain
what's going on with '@unknown' to someone who hasn't read the Swift 5
release notes.

jrose-apple commented:
(yes, I know it needs rebasing)

jrose-apple committed:
Don't require '@unknown' in playgrounds or in the debugger

We still model the enums as non-exhaustive so that someone /can/
handle unknown cases, which may be important for imported enums. But
we won't diagnose a problem if the only missing case is '@unknown'.

jrose-apple committed:
Downgrade exhaustivity errors to warnings if '@unknown' is present

The other half of '@unknown' in Sema. Again, the diagnostics here
could be improved; rather than a generic "switch must be exhaustive",
it could say something about unknown case handling known cases.

One interesting detail here: '@unknown' is only supposed to match
/fully/ missing cases. If a case is /partly/ accounted for, not
handling the rest is still an error, even if an unknown case is
present.

This only works with switches over single enum values, not values that
contain an enum with unknown cases. That's coming in a later commit.
(It was easier to get this part working first.)

jrose-apple commented:
Rebased on top of the new implementation of frozen enums! (already in-tree)

@CodaFi, I'm sorry to subject you to such a diff, but I welcome your SpaceEngine wisdom once again.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
@swift-ci Please test compiler performance

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
Oops, the very last change I did is wrong.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
This fails on Linux, and I can reproduce the failure, but I have no idea why. It's emitting the fix-its in a different order, but I don't see anything that should be non-deterministic or implementation-defined.

jrose-apple commented:
…should have known it'd be the std::sort.

@swift-ci Please test

CodaFi commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -78,13 +78,22 @@ namespace {
       Disjunct,
-      BooleanConstant
+      BooleanConstant,
+      UnknownCase,
```
You wound up making this a space kind anyway?  For those that want context, we discussed this in private.  The idea then was that you could traverse the pattern space to look for these - like we're already doing to find downgradable spaces - rather than using a subtraction and flattening them out.

jrose-apple replied:
Ah, heh, I forgot about that conversation. I think I threw out most of it once we decided `unknown` could match patterns *containing* non-frozen enums as well, though. Do you still think I should take it out?

CodaFi replied:
The space traversal for downgradable patterns looks into nested patterns as well - unless I'm not quite understanding you here.  I haven't kept up with the Swift Evolution discussion around this all that well.

jrose-apple replied:
I'll take a look tomorrow. Given that the user side of `unknown` had to be handled specially anyway, what I think you're saying might make more sense.

jrose-apple replied:
…yeah, I'm not sure this is going to work. I can't figure out how to deal with this example that way:

```swift
/*non-frozen*/ public enum Test {
  case a, b, c
}

@inlinable public func consume(_ x: Test) {
  switch x {
  case .a, .b: break
  }
}
```

In this case the compiler should suggest adding both `case .c` and `unknown`. If you decompose the `Test` space, you lose the non-frozen-ness, but if you don't decompose you can't subtract out the `a` and `b` cases.

I think the other way worked when the only thing that could match future cases was irrefutable patterns, and so the suggestion for the above example could just be `default`. But now that we have to distinguish these situations I think an Unknown space does make sense.

jrose-apple committed:
[AST] Adjust the API of CaseLabelItem for `unknown`

Get these simple changes out of the way first. No functionality change.

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented:
@swift-ci Please smoke test

jrose-apple commented on lib/Sema/TypeCheckSwitchStmt.cpp:
```diff
@@ -440,7 +493,8 @@ namespace {
         PAIRCASE (SpaceKind::Disjunct, SpaceKind::Disjunct):
-        PAIRCASE (SpaceKind::BooleanConstant, SpaceKind::Disjunct): {
+        PAIRCASE (SpaceKind::BooleanConstant, SpaceKind::Disjunct):
+        PAIRCASE (SpaceKind::UnknownCase, SpaceKind::Disjunct): {
```
Yeah, I'd kinda like to do that for all the symmetrical ones. I didn't change that structure in this PR, though.

xedin replied:
Makes sense, just wanted to point it out!

xedin reviewed (approved):
Looks great, @jrose-apple! I think it makes sense to have `Unknown` as a type of space and check for it only after initial exhaustiveness check.

jrose-apple committed:
Add parsing support for `@unknown` (SE-0192)

This is our first statement attribute, made more complicated by the
fact that a 'case'/'default' isn't really a normal statement. I've
chosen /not/ to implement a general statement attribute logic like we
have for types and decls at this time, but I did get the compiler
parsing arbitrary attributes before 'case' and 'default'. As a bonus,
we now treat all cases within functions as being switch-like rather
than enum-like, which is better for recovery when not in a switch.

jrose-apple committed:
[test] '@unknown' should be ignored post-Sema

jrose-apple committed:
'@unknown' can match unknown cases in nested positions

That is, when matching non-frozen enums at non-top-level positions:

    switch (nonFrozenEnum1, nonFrozenEnum2) {
    case (.singleKnownCase1, .singleKnownCase2): ...
    unknown: ...
    }

...it's sufficient to use '@unknown' to match

  (.singleKnownCase1, .someFutureCase2)
  (.someFutureCase1, .singleKnownCase2)
  (.someFutureCase1, .someFutureCase2)

jrose-apple committed:
Allow '@unknown' to match any enums not explicitly marked frozen

If a client wants to defend themselves against new cases in libraries
they use, or even in their own code, they're allowed to.

jrose-apple committed:
Clean up fix-it generation for missing cases, esp. with '@unknown'

- Combine the common logic for editor mode and non-editor mode.
- Do a better job minimizing fix-its.
- If '@unknown' is the only missing case, put `fatalError()` in the
  Xcode placeholder, since that's what the compiler would have done.

jrose-apple committed:
Enforce restrictions on '@unknown'

- Must be a single pattern
- Must be last in the switch
- Must not have a 'where' clause
- Must specifically be the "any" pattern if using 'case'

We may lift some of these restrictions in the future, but that would
need a new proposal.

jrose-apple committed:
Prefer `@unknown default` over `@unknown case _` in diagnostics

(and fix-its)

jrose-apple commented:
Okay, final version is ready. None of the space engine stuff changed in any interesting ways, but it may be worth looking over the rest of it again, @xedin.

@swift-ci Please test

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test source compatibility

jrose-apple commented:
@swift-ci Please test

rintaro commented on utils/gyb_syntax_support/StmtNodes.py:
```diff
@@ -273,6 +274,13 @@
+         children=[
+             Child('UnknownKeyword', kind='Token'),
+             Child('Colon', kind='ColonToken'),
+         ]),
```
This is not used. Can I remove this?

jrose-apple replied:
Whoops, I thought I took it out already. Yes, sorry!
